crates/red_knot_python_semantic/src/types/type_api.rs

```diff
@@ -0,0 +1,119 @@
+use crate::types::{Class, IntersectionBuilder, Type};
+use crate::Db;
+
+use thiserror::Error;
+
+#[derive(Debug, Error)]
+pub(crate) enum TypeApiError {
+    #[error("Wrong number of arguments")]
+    WrongArity,
+    #[error("Unknown API expression")]
+    UnknownApiExpression,
+    #[error("Failed assertion")]
+    FailedAssertion,
+}
+
+type Result<T> = std::result::Result<T, TypeApiError>;
+
+fn expect_one_argument<'db>(mut arguments: impl Iterator<Item = Type<'db>>) -> Result<Type<'db>> {
+    let first = arguments.next().ok_or(TypeApiError::WrongArity)?;
+    if arguments.next().is_some() {
+        return Err(TypeApiError::WrongArity);
+    }
+    Ok(first)
+}
+
+fn expect_two_arguments<'db>(
+    mut arguments: impl Iterator<Item = Type<'db>>,
+) -> Result<(Type<'db>, Type<'db>)> {
+    let first = arguments.next().ok_or(TypeApiError::WrongArity)?;
+    let second = arguments.next().ok_or(TypeApiError::WrongArity)?;
+    if arguments.next().is_some() {
+        return Err(TypeApiError::WrongArity);
+    }
+    Ok((first, second))
+}
+
+pub(crate) fn resolve_type_operation<'db>(
+    db: &'db dyn Db,
+    class: Class<'db>,
+    arguments: impl Iterator<Item = Type<'db>>,
+) -> Result<Type<'db>> {
+    match class.name(db).as_str() {
+        "Negate" => {
+            let ty = expect_one_argument(arguments)?;
+            Ok(ty.negate(db))
+        }
```

Comment:
nit: in most proposals for this feature (e.g. https://github.com/python/typing/issues/801), this is generally referred to as a `Not` type rather than a `Negate` type, so that would be the name I'd be most familiar with for this